pkg/operator/certrotation/client_cert_rotation_controller.go

```diff
@@ -57,32 +50,37 @@ func NewCertRotationController(
 	caBundleRotation CABundleRotation,
 	targetRotation TargetRotation,
 	operatorClient v1helpers.StaticPodOperatorClient,
-) (*CertRotationController, error) {
+	recorder events.Recorder,
+) factory.Controller {
 	c := &CertRotationController{
-		name: name,
-
+		name:             name,
 		SigningRotation:  signingRotation,
 		CABundleRotation: caBundleRotation,
 		TargetRotation:   targetRotation,
 		OperatorClient:   operatorClient,
-
-		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), strings.Replace(name, "-", "_", -1)),
 	}
-
-	signingRotation.Informer.Informer().AddEventHandler(c.eventHandler())
-	caBundleRotation.Informer.Informer().AddEventHandler(c.eventHandler())
-	targetRotation.Informer.Informer().AddEventHandler(c.eventHandler())
-
-	c.cachesToSync = append(c.cachesToSync, signingRotation.Informer.Informer().HasSynced)
-	c.cachesToSync = append(c.cachesToSync, caBundleRotation.Informer.Informer().HasSynced)
-	c.cachesToSync = append(c.cachesToSync, targetRotation.Informer.Informer().HasSynced)
-
-	return c, nil
+	return factory.New().
+		ResyncEvery(30*time.Second).
+		WithSync(c.Sync).
+		WithInformers(
+			signingRotation.Informer.Informer(),
+			caBundleRotation.Informer.Informer(),
+			targetRotation.Informer.Informer(),
+		).
+		WithPostRunHooks(
+			c.targetCertRecheckerPostRunHook,
+		).
+		ToController("CertRotationController", recorder.WithComponentSuffix("cert-rotation-controller"))
 }
 
-func (c CertRotationController) sync() error {
+func (c CertRotationController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
why is this a non-pointer now?

Reply:
it does not mutate the CertRotationController and still satisfy the interface.

Reply:
I hate this in Go that call-by-value receivers semantics is misused for read-only enforcement. It is slow to copy around huge objects every time.